app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt

```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
     }
 
     override fun onPlanMyJourneyClicked() {
-        val apiService = JourneyPlannerApiService.createApiService()
-
-        call = apiService.getJourneyResults(
+        call = JourneyPlannerApiService.createApiService().getJourneyResults(
             view.getFromLocation().text.toString(),
             view.getToLocation().text.toString()
         )
 
-/*            TODO("Default adapter treats response code 300 Multiple Options as unsuccessful and")
-            TODO("puts response in the errorBody but is actually a valid response.")
-            TODO("Implement custom adapter to parse the valid response in the errorBody")   */
-        call!!.enqueue(object : Callback<JourneyPlannerDisambiguationResult> {
-            override fun onFailure(call: Call<JourneyPlannerDisambiguationResult>, throwable: Throwable) {
-                Log.e(TAG, "Other error: ${throwable.message}")
-                Log.e(TAG, "Other error: ${throwable.stackTrace.forEach { println(it) }}")
+        call!!.enqueue(object : Callback<JourneyPlanner.ItineraryResult> {
```

Comment:
presenters are about usecase->view interaction.

You're mixing the usecase layer here. Better is to ask for data to the usecase, and the use case deals with the api and all its particularities. Presenters just care about wiring logic

Reply:
Added uses cases but I'm not convinced with my implementation

Reply:
No, you need something of the sort of a Repository / Provider / etc. I'll give you some ideas of what to do later

Reply:
Added repository, datasource and use cases/interactors.

Reply:
Fixed repository which was just delegating to datasource and removed latter